src/vm/types/mod.rs

```diff
@@ -0,0 +1,450 @@
+
+impl PartialEq for TupleData {
+    fn eq(&self, other: &TupleData) -> bool {
+        self.data_map == other.data_map
+    }
+}
+
+impl Hash for TupleData {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.data_map.hash(state);
+    }
+}
+
+pub const NONE: Value = Value::Optional(OptionalData { data: None });
+
+impl Value {
+    pub fn some(data: Value) -> Value {
+        Value::Optional(OptionalData {
+            data: Some(Box::new(data)) })
+    }
+
+    pub fn none() -> Value {
+        Value::Optional(OptionalData {
+            data: None })
+    }
+
+    pub fn okay(data: Value) -> Value {
+        Value::Response(ResponseData { 
+            committed: true,
+            data: Box::new(data) })
+    }
+
+    pub fn error(data: Value) -> Value {
+        Value::Response(ResponseData { 
+            committed: false,
+            data: Box::new(data) })
+    }
+
+    /// Invariant: the supplied Values have already been "checked", i.e., it's a valid Value object
+    ///  this invariant is enforced through the Value constructors, each of which checks to ensure
+    ///  that any typing data is correct.
+    pub fn list_with_type(list_data: Vec<Value>, expected_type: ListTypeData) -> Result<Value> {
+        if expected_type.size()? > MAX_VALUE_SIZE {
+            return Err(RuntimeErrorType::ValueTooLarge.into())
+        }
+
+        if (expected_type.max_len as usize) < list_data.len() {
+            return Err(InterpreterError::FailureConstructingListWithType.into())
+        }
+
+        let expected_item_type = expected_type.get_list_item_type();
+
+        for item in &list_data {
+            if !expected_item_type.admits(&item) {
+                return Err(InterpreterError::FailureConstructingListWithType.into())
+            }
+        }
+
+        Ok(Value::List(ListData { data: list_data, type_signature: expected_type }))
+    }
+
+    pub fn list_from(list_data: Vec<Value>) -> Result<Value> {
+        let type_sig = TypeSignature::construct_parent_list_type(&list_data)?;
+        // Aaron: at this point, we've _already_ allocated memory for this type.
+        //     (e.g., from a (map...) call, or a (list...) call.
+        //     this is a problem _if_ the static analyzer cannot already prevent
+        //     this case. This applies to all the constructor size checks.
+        if type_sig.size()? > MAX_VALUE_SIZE {
+            return Err(RuntimeErrorType::ValueTooLarge.into())
+        }
+        Ok(Value::List(ListData { data: list_data, type_signature: type_sig }))
+    }
+
+    pub fn buff_from(buff_data: Vec<u8>) -> Result<Value> {
+        if buff_data.len() > u32::max_value() as usize {
+            Err(RuntimeErrorType::BufferTooLarge.into())
+        } else if buff_data.len() as i128 > MAX_VALUE_SIZE {
+            Err(RuntimeErrorType::ValueTooLarge.into())
+        } else {
+            Ok(Value::Buffer(BuffData { data: buff_data }))
+        }
+    }
+
+    pub fn size(&self) -> Result<i128> {
```

Comment:
How can this method fail?  What does it mean for it to fail?

Reply:
Similar to the u32 max value check -- type sizes should _never_ result in arithmetic overflows, but that's a property of the MAX_VALUE_SIZE check, not the Rust types. Furthermore, in the value bounding branch, `.size()` is the only method that is allowed to be invoked during TypeSignature construction (meaning its the only function which may be called on a too-large type), and is used for the actual MAX_VALUE_SIZE enforcement, meaning that some code _could_ conceivably result in an arithmetic overflow in size().

Reply:
I feel like this is a risky approach.  I think it might be prudent to make the TypeSignature construction itself fail if the size would be too big (thereby making such type unrepresentable), and in so doing, deny the instantiation of Clarity code that could produce unrepresentable types.  What are your thoughts?

Reply:
That's exactly what it will do -- type signature constructors are all guarded. However, rather than implement size computation twice, the constructors call `size` to check whether or not the size of the type is too large -- this means that the size function _needs_ to be able to handle the case in which a type signature construction will fail. As for the size() invocation in Value, it should never fail, and this can be enforced with an `expects()`.